File: dotnet/src/RocketMQ.Client/Client/Latency.cs
using System.Collections.Generic;
using System.Threading;
using RocketMQ.Common;

namespace RocketMQ.Client;

/// <summary>
/// 发送延迟故障容错（对应 org.apache.rocketmq.client.latency.* 与 Python client/latency.py）。
///
/// 实现 MQFaultStrategy + LatencyFaultToleranceImpl（带 FaultItem）：追踪每个 broker 的
/// 发送延迟，延迟过高或发生异常时<b>隔离</b>一段时间（不分配给新消息），默认关闭。
///
/// 与 Java 关键点逐条对齐：
///   * latencyMax / notAvailableDuration 两套阈值表（逐字一致）；
///   * UpdateFaultItem 的 notAvailableDuration 取 ComputeNotAvailableDuration，
///     隔离（异常）场景固定按 10000ms 算档位；
///   * FaultItem.IsAvailable() = now &gt;= startTimestamp（隔离期未过则不可用）；
///   * LatencyFaultToleranceImpl.IsAvailable/IsReachable 在没有记录时返回 true，
///     即"从未出过问题的 broker 默认可用/可达"。
///
/// 有意差异（与 Python 参考实现一致）：省略 Java 的"后台可达性探测线程"（startDetector），
/// reachable 由 UpdateFaultItem 的调用方写入。默认 sendLatencyFaultEnable=false，开启后才记录与生效。
/// </summary>
public class FaultItem
{
    public FaultItem(string name)
    {
        Name = name;
    }

    public string Name { get; }

    public double CurrentLatency { get; set; }

    public long StartTimestamp { get; private set; }

    /// <summary>Java：only when now + dur &gt; startTimestamp 才更新（保持最长隔离期）。</summary>
    public void UpdateNotAvailableDuration(long notAvailableDurationMillis)
    {
        long now = UtilAll.CurrentTimeMillis();
        if (notAvailableDurationMillis > 0 && now + notAvailableDurationMillis > StartTimestamp)
        {
            StartTimestamp = now + notAvailableDurationMillis;
        }
    }

    public bool IsAvailable() => UtilAll.CurrentTimeMillis() >= StartTimestamp;

    public bool IsReachable() { return _reachableFlag; }

    public void SetReachable(bool reachable) { _reachableFlag = reachable; }

    private bool _reachableFlag = true;
}

/// <summary>
/// 对应 Java client.latency.LatencyFaultToleranceImpl（纯内存版，无探测线程）。
/// 表用锁保护；FaultItem 字段由表锁串行化访问。
/// </summary>
public class LatencyFaultToleranceImpl
{
    private readonly object _lock = new();
    private readonly Dictionary<string, FaultItem> _faultItemTable = new();

    public void UpdateFaultItem(string name, double currentLatency,
                                long notAvailableDurationMillis, bool reachable)
    {
        lock (_lock)
        {
            if (!_faultItemTable.TryGetValue(name, out FaultItem? item))
            {
                item = new FaultItem(name);
                _faultItemTable[name] = item;
            }
            item.CurrentLatency = currentLatency;
            item.UpdateNotAvailableDuration(notAvailableDurationMillis);
            item.SetReachable(reachable);
        }
    }

    public bool IsAvailable(string name)
    {
        FaultItem? item;
        lock (_lock)
        {
            _faultItemTable.TryGetValue(name, out item);
        }

        // 没有记录 = 从未出过问题，默认可用（与 Java/Python 一致）
        return item == null || item.IsAvailable();
    }

    public bool IsReachable(string name)
    {
        FaultItem? item;
        lock (_lock)
        {
            _faultItemTable.TryGetValue(name, out item);
        }

        return item == null || item.IsReachable();
    }

    public void Remove(string name)
    {
        lock (_lock)
        {
            _faultItemTable.Remove(name);
        }
    }

    /// <summary>供测试/诊断读取（不存在返回 null）。</summary>
    public FaultItem? GetFaultItem(string name)
    {
        lock (_lock)
        {
            return _faultItemTable.TryGetValue(name, out FaultItem? item) ? item : null;
        }
    }
}

/// <summary>
/// 对应 Java client.latency.MQFaultStrategy。
///
/// 仅当 SendLatencyFaultEnable 为 true 时，发送选队列阶段会：
///   1) 优先选 available（隔离期已过）的 broker；
///   2) 否则选 reachable 的 broker；
///   3) 否则退化为普通轮询。
/// 发送结果/异常会回调 UpdateFaultItem 写延迟与隔离信息。
/// </summary>
public class MQFaultStrategy
{
    // 两套阈值表（Java DefaultMQProducer 的默认值，逐字一致）
    public static readonly int[] LatencyMax = { 50, 100, 550, 1800, 3000, 5000, 15000 };
    public static readonly int[] NotAvailableDuration = { 0, 0, 2000, 5000, 6000, 10000, 30000 };

    private bool _sendLatencyFaultEnable;

    public MQFaultStrategy(bool sendLatencyFaultEnable = false)
    {
        _sendLatencyFaultEnable = sendLatencyFaultEnable;
    }

    public bool IsSendLatencyFaultEnable() => _sendLatencyFaultEnable;

    public void SetSendLatencyFaultEnable(bool enable) { _sendLatencyFaultEnable = enable; }

    public LatencyFaultToleranceImpl LatencyFaultTolerance { get; } = new();

    /// <summary>
    /// 队列选择（对应 Python select_one_message_queue / Java 同名方法）。
    /// lastBrokerName 非空时尽量避开该 broker；enable=false 时退化为普通轮询。
    /// </summary>
    public MessageQueue SelectOneMessageQueue(TopicPublishInfo tpInfo,
                                              string? lastBrokerName,
                                              bool resetIndex = false)
    {
        if (_sendLatencyFaultEnable)
        {
            if (resetIndex)
            {
                tpInfo.ResetIndex();
            }

            // 1) available：隔离期已过的 broker
            MessageQueue? mq = tpInfo.SelectOneMessageQueue(
                q => LatencyFaultTolerance.IsAvailable(q.BrokerName),
                q => lastBrokerName == null || q.BrokerName != lastBrokerName);
            if (mq is not null)
            {
                return mq;
            }

            // 2) reachable：隔离中但通道仍可达的 broker
            mq = tpInfo.SelectOneMessageQueue(
                q => LatencyFaultTolerance.IsReachable(q.BrokerName),
                q => lastBrokerName == null || q.BrokerName != lastBrokerName);
            if (mq is not null)
            {
                return mq;
            }

            // 3) 全部被隔离且不可达：退化为普通轮询
            return tpInfo.SelectOneMessageQueue();
        }

        // 关闭时退化为普通轮询（lastBrokerName 非空时避开它；全部同名时该重载内部已兜底）
        return tpInfo.SelectOneMessageQueue(lastBrokerName ?? "");
    }

    /// <summary>
    /// 故障记录：isolation=true 时 latency 固定按 10000ms 算档位（→ 隔离 10000ms）。
    /// 未开启时直接忽略（与 Java/Python 一致）。
    /// </summary>
    public void UpdateFaultItem(string brokerName, double currentLatency,
                                bool isolation, bool reachable)
    {
        if (!_sendLatencyFaultEnable)
        {
            return;
        }

        double latency = isolation ? 10000.0 : currentLatency;
        long duration = ComputeNotAvailableDuration(latency);
        LatencyFaultTolerance.UpdateFaultItem(brokerName, currentLatency, duration, reachable);
    }

    /// <summary>从表尾向首找第一个 latency &gt;= LatencyMax[i] 的档位；都不满足返回 0。</summary>
    public long ComputeNotAvailableDuration(double currentLatency)
    {
        for (int i = LatencyMax.Length - 1; i >= 0; --i)
        {
            if (currentLatency >= LatencyMax[i])
            {
                return NotAvailableDuration[i];
            }
        }
        return 0;
    }
}

File: python/rocketmq/client/latency.py
# -*- coding: utf-8 -*-
"""发送延迟故障容错（对应 org.apache.rocketmq.client.latency.*）。

实现 MQFaultStrategy + LatencyFaultToleranceImpl（带 FaultItem）：追踪每个 broker 的
发送延迟，延迟过高或发生异常时**隔离**一段时间（不分配给新消息），默认关闭。

与 Java 关键点逐条对齐：
  * latencyMax / notAvailableDuration 两套阈值表；
  * updateFaultItem 的 notAvailableDuration 取 ``computeNotAvailableDuration``，
    隔离（异常）场景固定按 10000ms 算档位；
  * FaultItem.isAvailable() = now >= startTimestamp（隔离期未过则不可用）；
  * LatencyFaultToleranceImpl.isAvailable/isReachable 在没有记录时返回 true，
    即"从未出过问题的 broker 默认可用/可达"。

默认 sendLatencyFaultEnable=false，与 Java 一致；开启后才会记录与生效。
"""
from __future__ import annotations

import threading
import time
from typing import Callable, Dict, Optional

from ..common.message import MessageQueue


class FaultItem:
    """单个 broker 的故障项（对应 Java LatencyFaultToleranceImpl.FaultItem）。"""

    def __init__(self, name: str):
        self.name = name
        self.current_latency: float = 0.0
        self.start_timestamp: float = 0.0
        self.check_stamp: float = 0.0
        self.reachable_flag: bool = True

    def update_not_available_duration(self, not_available_duration: float) -> None:
        # Java：only when now + dur > startTimestamp 才更新（保持最长隔离期）
        if not_available_duration > 0 and time.time() * 1000.0 + not_available_duration > self.start_timestamp:
            self.start_timestamp = time.time() * 1000.0 + not_available_duration

    def is_available(self) -> bool:
        return time.time() * 1000.0 >= self.start_timestamp

    def is_reachable(self) -> bool:
        return self.reachable_flag

    def __repr__(self):
        return "FaultItem{name=%s, latency=%.0f, startTs=%.0f, reachable=%s}" % (
            self.name, self.current_latency, self.start_timestamp, self.reachable_flag)


class LatencyFaultToleranceImpl:
    """对应 Java client.latency.LatencyFaultToleranceImpl（简化为纯内存版，无探测线程）。

    省略 Java 的"后台可达性探测线程"（startDetector），因为探测依赖真实 broker 连接；
    本地保留 reachableFlag 语义：updateFaultItem(..., reachable) 时写 reachableFlag。
    """

    def __init__(self):
        self._fault_item_table: Dict[str, FaultItem] = {}
        self._lock = threading.RLock()

    def update_fault_item(self, name: str, current_latency: float,
                          not_available_duration: float, reachable: bool) -> None:
        with self._lock:
            item = self._fault_item_table.get(name)
            if item is None:
                item = FaultItem(name)
                item.current_latency = current_latency
                item.update_not_available_duration(not_available_duration)
                item.reachable_flag = reachable
                self._fault_item_table[name] = item
                return
            item.current_latency = current_latency
            item.update_not_available_duration(not_available_duration)
            item.reachable_flag = reachable

    def is_available(self, name: str) -> bool:
        with self._lock:
            item = self._fault_item_table.get(name)
        if item is not None:
            return item.is_available()
        return True

    def is_reachable(self, name: str) -> bool:
        with self._lock:
            item = self._fault_item_table.get(name)
        if item is not None:
            return item.is_reachable()
        return True

    def remove(self, name: str) -> None:
        with self._lock:
            self._fault_item_table.pop(name, None)

    def get_fault_item(self, name: str) -> Optional[FaultItem]:
        with self._lock:
            return self._fault_item_table.get(name)


class MQFaultStrategy:
    """对应 Java client.latency.MQFaultStrategy。

    仅当 ``send_latency_fault_enable`` 为 True 时，发送选队列阶段会：
      1) 优先选 available（隔离期已过）的 broker；
      2) 否则选 reachable 的 broker；
      3) 否则退化为普通轮询。
    发送结果/异常会回调 ``update_fault_item`` 写延迟与隔离信息。
    """

    LATENCY_MAX = [50, 100, 550, 1800, 3000, 5000, 15000]
    NOT_AVAILABLE_DURATION = [0, 0, 2000, 5000, 6000, 10000, 30000]

    def __init__(self, send_latency_fault_enable: bool = False):
        self._send_latency_fault_enable = send_latency_fault_enable
        self._latency_fault_tolerance = LatencyFaultToleranceImpl()
        self.latency_max = list(self.LATENCY_MAX)
        self.not_available_duration = list(self.NOT_AVAILABLE_DURATION)

    # ---- 配置 ----
    def is_send_latency_fault_enable(self) -> bool:
        return self._send_latency_fault_enable

    def set_send_latency_fault_enable(self, enable: bool) -> None:
        self._send_latency_fault_enable = enable

    # ---- 队列选择 ----
    def _available_filter(self, mq: MessageQueue) -> bool:
        return self._latency_fault_tolerance.is_available(mq.get_broker_name())

    def _reachable_filter(self, mq: MessageQueue) -> bool:
        return self._latency_fault_tolerance.is_reachable(mq.get_broker_name())

    def select_one_message_queue(self, tp_info, last_broker_name: Optional[str],
                                 reset_index: bool = False) -> MessageQueue:
        broker_filter: Callable[[MessageQueue], bool] = (
            lambda mq: last_broker_name is None or mq.get_broker_name() != last_broker_name)
        if self._send_latency_fault_enable:
            if reset_index:
                tp_info.reset_index()
            mq = tp_info.select_one_message_queue(self._available_filter, broker_filter)
            if mq is not None:
                return mq
            mq = tp_info.select_one_message_queue(self._reachable_filter, broker_filter)
            if mq is not None:
                return mq
            return tp_info.select_one_message_queue()
        mq = tp_info.select_one_message_queue(broker_filter)
        if mq is not None:
            return mq
        return tp_info.select_one_message_queue()

    # ---- 故障记录 ----
    def update_fault_item(self, broker_name: str, current_latency: float,
                          isolation: bool, reachable: bool) -> None:
        if not self._send_latency_fault_enable:
            return
        latency = 10000 if isolation else current_latency
        duration = self._compute_not_available_duration(latency)
        self._latency_fault_tolerance.update_fault_item(broker_name, current_latency, duration, reachable)

    def _compute_not_available_duration(self, current_latency: float) -> float:
        for i in range(len(self.latency_max) - 1, -1, -1):
            if current_latency >= self.latency_max[i]:
                return self.not_available_duration[i]
        return 0

    @property
    def latency_fault_tolerance(self) -> LatencyFaultToleranceImpl:
        return self._latency_fault_tolerance


__all__ = ["FaultItem", "LatencyFaultToleranceImpl", "MQFaultStrategy"]
